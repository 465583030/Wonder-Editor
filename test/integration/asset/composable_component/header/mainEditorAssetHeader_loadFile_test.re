open Wonder_jest;

open AssetTreeTwoLayerTypeTool;

open Expect;

open Expect.Operators;

open Sinon;

open AssetTreeNodeType;

let _ =
  describe("MainEditorAssetHeader", () => {
    let sandbox = getSandboxDefaultVal();

    beforeEach(() => {
      sandbox := createSandbox();

      MainEditorSceneTool.initState(~sandbox, ());
      MainEditorSceneTool.createDefaultScene(
        sandbox,
        MainEditorAssetTool.initAssetTree,
      );

      EventListenerTool.buildFakeDom()
      |> EventListenerTool.stubGetElementByIdReturnFakeDom;
    });

    afterEach(() => {
      restoreSandbox(refJsObjToSandbox(sandbox^));
      StateEditorService.getState()
      |> AssetCurrentNodeDataEditorService.clearCurrentNodeData
      |> AssetCurrentNodeParentIdEditorService.clearCurrentNodeParentId
      |> StateEditorService.setState
      |> ignore;
    });

    describe("test load file", () => {
      beforeEach(() => {
        MainEditorAssetTool.buildFakeFileReader();
        MainEditorAssetTool.buildFakeImage();
      });
      describe("test snapshot", () =>
        describe("if not select specific treeNode", () =>
          testPromise("load file should add into root node children", () => {
            MainEditorAssetTool.buildTwoLayerAssetTreeRoot() |> ignore;

            MainEditorAssetTool.fileLoad(
              TestTool.getDispatch(),
              BaseEventTool.buildFileEvent(),
            )
            |> Js.Promise.then_(_ =>
                 BuildComponentTool.buildAssetComponent()
                 |> ReactTestTool.createSnapshotAndMatch
                 |> Js.Promise.resolve
               );
          })
        )
      );

      describe("test logic", () => {
        describe("test should add into root node children", () =>
          testPromise("test children node length", () => {
            MainEditorAssetTool.buildTwoLayerAssetTreeRoot() |> ignore;
            let uploadFileLength = 2;
            let originChildrenLen =
              StateEditorService.getState()
              |> AssetTreeRootEditorService.unsafeGetAssetTreeRoot
              |> (root => root.children)
              |> Js.Array.length;

            MainEditorAssetTool.fileLoad(
              TestTool.getDispatch(),
              BaseEventTool.buildFileEvent(),
            )
            |> Js.Promise.then_(_ =>
                 StateEditorService.getState()
                 |> AssetTreeRootEditorService.unsafeGetAssetTreeRoot
                 |> (root => root.children)
                 |> Js.Array.length
                 |> (lastLen => lastLen - originChildrenLen)
                 |> expect == uploadFileLength
                 |> Js.Promise.resolve
               );
          })
        );

        describe("test should add into nodeMap", () => {
          describe("test imageBase64Map", () => {
            testPromise("add image base64 to imageBase64Map", () => {
              let assetTreeDomRecord =
                MainEditorAssetTool.buildTwoLayerAssetTreeRoot();
              let imgBase64 = "newImgBase64";

              MainEditorAssetTool.fileLoad(
                TestTool.getDispatch(),
                BaseEventTool.buildFileEvent(~imgSrc=imgBase64, ()),
              )
              |> Js.Promise.then_(_ => {
                   assetTreeDomRecord
                   |> MainEditorAssetNodeTool.OperateTwoLayer.getUploadedeTextureNodeDomIndex
                   |> MainEditorAssetTool.clickAssetChildrenNodeToSetCurrentNode;

                   StateEditorService.getState()
                   |> AssetImageBase64MapEditorService.getImageBase64Map
                   |> WonderCommonlib.SparseMapService.unsafeGet(
                        MainEditorAssetNodeTool.getTextureIndexFromCurrentNodeId(),
                      )
                   |> expect == imgBase64
                   |> Js.Promise.resolve;
                 });
            });
            testPromise(
              "test show texture image, get it base64 from imageBase64Map", () => {
              MainEditorAssetTool.buildTwoLayerAssetTreeRoot() |> ignore;
              let imgBase64 = "newImgBase64";

              MainEditorAssetTool.fileLoad(
                TestTool.getDispatch(),
                BaseEventTool.buildFileEvent(~imgSrc=imgBase64, ()),
              )
              |> Js.Promise.then_(_ =>
                   BuildComponentTool.buildAssetComponent()
                   |> ReactTestTool.createSnapshotAndMatch
                   |> Js.Promise.resolve
                 );
            });
          });

          describe("test textureNodeMap", () =>
            testPromise("add created texture index to textureNodeMap", () => {
              let assetTreeDomRecord =
                MainEditorAssetTool.buildTwoLayerAssetTreeRoot();

              MainEditorAssetTool.fileLoad(
                TestTool.getDispatch(),
                BaseEventTool.buildFileEvent(),
              )
              |> Js.Promise.then_(_ => {
                   assetTreeDomRecord
                   |> MainEditorAssetNodeTool.OperateTwoLayer.getUploadedeTextureNodeDomIndex
                   |> MainEditorAssetTool.clickAssetChildrenNodeToSetCurrentNode;

                   MainEditorAssetNodeTool.getTextureIndexFromCurrentNodeId()
                   |>
                   expect == MainEditorAssetNodeTool.OperateTwoLayer.getUploadedTextureIndex(
                               assetTreeDomRecord,
                             )
                   |> Js.Promise.resolve;
                 });
            })
          );

          describe("test jsonNodeMap", () =>
            testPromise("add json string to jsonNodeMap", () => {
              let assetTreeDomRecord =
                MainEditorAssetTool.buildTwoLayerAssetTreeRoot();
              let jsonName = "newLoadJson.json";
              let jsonResult = "I'm the result";

              MainEditorAssetTool.fileLoad(
                TestTool.getDispatch(),
                BaseEventTool.buildFileEvent(~jsonName, ~jsonResult, ()),
              )
              |> Js.Promise.then_(_ => {
                   assetTreeDomRecord
                   |> MainEditorAssetNodeTool.OperateTwoLayer.getUploadedeJsonNodeDomIndex
                   |> MainEditorAssetTool.clickAssetChildrenNodeToSetCurrentNode;

                   let {name, jsonResult}: AssetNodeType.jsonResultType =
                     StateEditorService.getState()
                     |> AssetJsonNodeMapEditorService.getJsonNodeMap
                     |> WonderCommonlib.SparseMapService.unsafeGet(
                          MainEditorAssetNodeTool.getCurrentNodeId(),
                        );

                   (name, jsonResult)
                   |> expect == (jsonName, jsonResult)
                   |> Js.Promise.resolve;
                 });
            })
          );

          describe("test geometryNodeMap", () =>
            testPromise("add geometryIndex to geometryNodeMap", () => {
              let assetTreeDomRecord =
                MainEditorAssetTool.buildTwoLayerAssetTreeRoot();
              let jsonName = "newLoadJson.json";
              let jsonResult = "I'm the result";

              MainEditorAssetTool.fileLoad(
                TestTool.getDispatch(),
                BaseEventTool.buildFileEvent(~jsonName, ~jsonResult, ()),
              )
              |> Js.Promise.then_(_ => {
                   assetTreeDomRecord
                   |> MainEditorAssetNodeTool.OperateTwoLayer.getUploadedeJsonNodeDomIndex
                   |> MainEditorAssetTool.clickAssetChildrenNodeToSetCurrentNode;

                   let {name, jsonResult}: AssetNodeType.jsonResultType =
                     StateEditorService.getState()
                     |> AssetJsonNodeMapEditorService.getJsonNodeMap
                     |> WonderCommonlib.SparseMapService.unsafeGet(
                          MainEditorAssetNodeTool.getCurrentNodeId(),
                        );

                   (name, jsonResult)
                   |> expect == (jsonName, jsonResult)
                   |> Js.Promise.resolve;
                 });
            })
          );

          describe("test wdbNodeMap", () => {
            let _buildWDBPath = wdbName =>
              Node.Path.join([|
                Node.Process.cwd(),
                "./test/res/",
                {j|wdb/$wdbName.wdb|j},
              |]);
            let _getWDBArrayBuffer = wdbName => NodeExtendTool.readFileBufferSync(
                                                  _buildWDBPath(wdbName),
                                                )##buffer;

            beforeEach(() => {
              MainEditorAssetTool.buildFakeTextDecoder(
                MainEditorAssetTool.convertUint8ArrayToBuffer,
              );
              MainEditorAssetTool.buildFakeURL(sandbox^);
            });

            testPromise(
              "add wdbGameObject and wdbArrayBuffer to wdbNodeMap", () => {
              let assetTreeDomRecord =
                MainEditorAssetTool.buildTwoLayerAssetTreeRoot();
              let fileName = "BoxTextured";

              WonderLog.Log.print("start") |> ignore;
              MainEditorAssetTool.fileLoad(
                TestTool.getDispatch(),
                BaseEventTool.buildWdbFileEvent(
                  fileName,
                  _getWDBArrayBuffer(fileName),
                ),
              )
              |> Js.Promise.then_(_ => {
                   WonderLog.Log.print("asd") |> ignore;
                   let name = fileName;
                   /* assetTreeDomRecord
                      |> MainEditorAssetNodeTool.OperateTwoLayer.getUploadedeJsonNodeDomIndex
                      |> MainEditorAssetTool.clickAssetChildrenNodeToSetCurrentNode;

                      let {name, jsonResult}: AssetNodeType.jsonResultType =
                        StateEditorService.getState()
                        |> AssetJsonNodeMapEditorService.getJsonNodeMap
                        |> WonderCommonlib.SparseMapService.unsafeGet(
                             MainEditorAssetNodeTool.getCurrentNodeId(),
                           ); */

                   name |> expect == fileName |> Js.Promise.resolve;
                 });
            });
          });
        });
      });

      describe("deal with specific case", () => {
        let _getErrorTypeFile = () =>
          AssetTreeNodeUtils.getUploadFileType("aaa.bb");

        test("if upload error file type, should error", () => {
          let component = BuildComponentTool.buildConsole();
          let errorStub =
            createMethodStub(sandbox^, ConsoleTool.console, "error");

          AssetTreeNodeUtils.handleSpecificFuncByType(
            _getErrorTypeFile(),
            (() => (), () => (), () => ()),
          );

          errorStub
          |> expect
          |> toCalledWith([|
               "\n  Error:\n\n  title\n  handleSpecificFuncByType\n\n  description\n  the load file type is error\n\n  reason\n  \n\n  solution\n  \n\n  params\n  \n\n   ",
             |]);
        });
      });
    });
  });