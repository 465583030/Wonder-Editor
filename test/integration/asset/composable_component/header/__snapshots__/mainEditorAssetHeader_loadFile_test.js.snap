// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainEditorAssetHeader->loadFile test load file test logic test should add into nodeMap test imageBase64Map test show texture image, get it base64 from imageBase64Map 1`] = `
<article
  className="wonder-asset-component"
>
  <div
    className="asset-tree"
  >
    <article
      className="wonder-asset-header"
    >
      <div
        className="header-item"
      >
        <button
          onClick={[Function]}
        >
          addFolder
        </button>
      </div>
      <div
        className="header-item"
      >
        <button
          disabled={true}
          onClick={[Function]}
        >
          remove
        </button>
      </div>
      <div
        className="header-item"
      >
        <input
          className="file-upload"
          multiple={true}
          onChange={[Function]}
          type="file"
        />
      </div>
      <div
        className=""
      >
        <button
          onClick={[Function]}
        >
          addMaterial
        </button>
      </div>
    </article>
    <article
      className="wonder-asset-assetTree"
    >
      <ul
        className="wonder-tree-node"
      >
        <li
          onClick={[Function]}
          style={
            Object {
              "background": "#c0c0c0",
            }
          }
        >
          <div
            className="item-ground"
            draggable={true}
            onDragEnter={[Function]}
            onDragLeave={[Function]}
            onDragOver={[Function]}
            onDrop={[Function]}
          />
          <img
            src="./public/img/12.jpg"
          />
          Assets
        </li>
        <ul
          className="wonder-tree-node"
          draggable={true}
          onDragEnd={[Function]}
          onDragStart={[Function]}
        >
          <li
            onClick={[Function]}
            style={
              Object {
                "border": "1px solid red",
              }
            }
          >
            <div
              className="item-ground"
              draggable={true}
              onDragEnter={[Function]}
              onDragLeave={[Function]}
              onDragOver={[Function]}
              onDrop={[Function]}
            />
            <img
              src="./public/img/12.jpg"
            />
            newFolder
          </li>
        </ul>
        <ul
          className="wonder-tree-node"
          draggable={true}
          onDragEnd={[Function]}
          onDragStart={[Function]}
        >
          <li
            onClick={[Function]}
            style={
              Object {
                "border": "1px solid red",
              }
            }
          >
            <div
              className="item-ground"
              draggable={true}
              onDragEnter={[Function]}
              onDragLeave={[Function]}
              onDragOver={[Function]}
              onDrop={[Function]}
            />
            <img
              src="./public/img/12.jpg"
            />
            newFolder 1
          </li>
        </ul>
      </ul>
    </article>
  </div>
  <article
    className="wonder-asset-assetChildren"
  >
    <article
      className="wonder-asset-folderBox"
      id="folder-4"
      style={
        Object {
          "border": "1px solid red",
        }
      }
    >
      <div
        className="item-ground"
        draggable={true}
        onDragEnd={[Function]}
        onDragEnter={[Function]}
        onDragLeave={[Function]}
        onDragOver={[Function]}
        onDragStart={[Function]}
        onDrop={[Function]}
      />
      <img
        src="./public/img/11.jpg"
      />
      <span
        className="item-text"
      >
        newFolder
      </span>
    </article>
    <article
      className="wonder-asset-folderBox"
      id="folder-5"
      style={
        Object {
          "border": "1px solid red",
        }
      }
    >
      <div
        className="item-ground"
        draggable={true}
        onDragEnd={[Function]}
        onDragEnter={[Function]}
        onDragLeave={[Function]}
        onDragOver={[Function]}
        onDragStart={[Function]}
        onDrop={[Function]}
      />
      <img
        src="./public/img/11.jpg"
      />
      <span
        className="item-text"
      >
        newFolder 1
      </span>
    </article>
    <article
      className="wonder-asset-fileBox "
      onClick={[Function]}
    >
      <img
        onDragStart={[Function]}
        src="texture3img"
      />
      <span
        className="item-text"
      >
        texture3
      </span>
    </article>
    <article
      className="wonder-asset-fileBox "
      onClick={[Function]}
    >
      <img
        onDragStart={[Function]}
        src="./public/img/12.jpg"
      />
      <span
        className="item-text"
      >
        json.json
      </span>
    </article>
    <article
      className="wonder-asset-fileBox "
      onClick={[Function]}
    >
      <img
        onDragStart={[Function]}
        src="texture5img"
      />
      <span
        className="item-text"
      >
        texture5
      </span>
    </article>
    <article
      className="wonder-asset-fileBox "
      onClick={[Function]}
    >
      <img
        onDragStart={[Function]}
        src="newImgBase64"
      />
      <span
        className="item-text"
      >
        loadImg
      </span>
    </article>
    <article
      className="wonder-asset-fileBox "
      onClick={[Function]}
    >
      <img
        onDragStart={[Function]}
        src="./public/img/12.jpg"
      />
      <span
        className="item-text"
      >
        loadJson.json
      </span>
    </article>
  </article>
</article>
`;

exports[`MainEditorAssetHeader->loadFile test load file test snapshot if not select specific treeNode load file should add into root node children 1`] = `
<article
  className="wonder-asset-component"
>
  <div
    className="asset-tree"
  >
    <article
      className="wonder-asset-header"
    >
      <div
        className="header-item"
      >
        <button
          onClick={[Function]}
        >
          addFolder
        </button>
      </div>
      <div
        className="header-item"
      >
        <button
          disabled={true}
          onClick={[Function]}
        >
          remove
        </button>
      </div>
      <div
        className="header-item"
      >
        <input
          className="file-upload"
          multiple={true}
          onChange={[Function]}
          type="file"
        />
      </div>
      <div
        className=""
      >
        <button
          onClick={[Function]}
        >
          addMaterial
        </button>
      </div>
    </article>
    <article
      className="wonder-asset-assetTree"
    >
      <ul
        className="wonder-tree-node"
      >
        <li
          onClick={[Function]}
          style={
            Object {
              "background": "#c0c0c0",
            }
          }
        >
          <div
            className="item-ground"
            draggable={true}
            onDragEnter={[Function]}
            onDragLeave={[Function]}
            onDragOver={[Function]}
            onDrop={[Function]}
          />
          <img
            src="./public/img/12.jpg"
          />
          Assets
        </li>
        <ul
          className="wonder-tree-node"
          draggable={true}
          onDragEnd={[Function]}
          onDragStart={[Function]}
        >
          <li
            onClick={[Function]}
            style={
              Object {
                "border": "1px solid red",
              }
            }
          >
            <div
              className="item-ground"
              draggable={true}
              onDragEnter={[Function]}
              onDragLeave={[Function]}
              onDragOver={[Function]}
              onDrop={[Function]}
            />
            <img
              src="./public/img/12.jpg"
            />
            newFolder
          </li>
        </ul>
        <ul
          className="wonder-tree-node"
          draggable={true}
          onDragEnd={[Function]}
          onDragStart={[Function]}
        >
          <li
            onClick={[Function]}
            style={
              Object {
                "border": "1px solid red",
              }
            }
          >
            <div
              className="item-ground"
              draggable={true}
              onDragEnter={[Function]}
              onDragLeave={[Function]}
              onDragOver={[Function]}
              onDrop={[Function]}
            />
            <img
              src="./public/img/12.jpg"
            />
            newFolder 1
          </li>
        </ul>
      </ul>
    </article>
  </div>
  <article
    className="wonder-asset-assetChildren"
  >
    <article
      className="wonder-asset-folderBox"
      id="folder-4"
      style={
        Object {
          "border": "1px solid red",
        }
      }
    >
      <div
        className="item-ground"
        draggable={true}
        onDragEnd={[Function]}
        onDragEnter={[Function]}
        onDragLeave={[Function]}
        onDragOver={[Function]}
        onDragStart={[Function]}
        onDrop={[Function]}
      />
      <img
        src="./public/img/11.jpg"
      />
      <span
        className="item-text"
      >
        newFolder
      </span>
    </article>
    <article
      className="wonder-asset-folderBox"
      id="folder-5"
      style={
        Object {
          "border": "1px solid red",
        }
      }
    >
      <div
        className="item-ground"
        draggable={true}
        onDragEnd={[Function]}
        onDragEnter={[Function]}
        onDragLeave={[Function]}
        onDragOver={[Function]}
        onDragStart={[Function]}
        onDrop={[Function]}
      />
      <img
        src="./public/img/11.jpg"
      />
      <span
        className="item-text"
      >
        newFolder 1
      </span>
    </article>
    <article
      className="wonder-asset-fileBox "
      onClick={[Function]}
    >
      <img
        onDragStart={[Function]}
        src="texture3img"
      />
      <span
        className="item-text"
      >
        texture3
      </span>
    </article>
    <article
      className="wonder-asset-fileBox "
      onClick={[Function]}
    >
      <img
        onDragStart={[Function]}
        src="./public/img/12.jpg"
      />
      <span
        className="item-text"
      >
        json.json
      </span>
    </article>
    <article
      className="wonder-asset-fileBox "
      onClick={[Function]}
    >
      <img
        onDragStart={[Function]}
        src="texture5img"
      />
      <span
        className="item-text"
      >
        texture5
      </span>
    </article>
    <article
      className="wonder-asset-fileBox "
      onClick={[Function]}
    >
      <img
        onDragStart={[Function]}
        src="newImgBase64"
      />
      <span
        className="item-text"
      >
        loadImg
      </span>
    </article>
    <article
      className="wonder-asset-fileBox "
      onClick={[Function]}
    >
      <img
        onDragStart={[Function]}
        src="./public/img/12.jpg"
      />
      <span
        className="item-text"
      >
        loadJson.json
      </span>
    </article>
  </article>
</article>
`;
